path_finding.asciidoc

```diff
@@ -0,0 +1,76 @@
+
+[[lngraph]]
+image::images/LNGraphJuly2021.png[]
+
+If the sender and recipient are connected to other well-connected nodes and have at least one channel with adequate capacity - there will be thousands of paths. The problem becomes: choosing the _best_ path that will succeed in payment delivery, out of a list of thousands of possible paths.
+
+==== Selecting the best path
+
+To select the "best" path, we have to first define what we mean by "best". There may be many different criteria such as:
+
+* Paths with enough capacity. Obviously if a path doesn't have enough capacity to route our payment, then it is not a suitable path
+
+* Paths with low fees. If we have several candidates, we may want to select ones with lower fees.
+
+* Paths with short timelocks. We may want to avoid locking our funds for too long and therefore select paths with shorter timelocks.
+
+All of these criteria may be desirable to some extent and selecting paths that are favorable across many dimensions is not an easy task. Optimization problems like this may be too complex to solve for the "best" solution, but often can be solved for some approximation of the optimal. Which is good news, because otherwise path finding would be an intractable problem.
+
+=== Fundamentals of path finding
+
+The problem of path finding in the Lightning Network falls under a general category of _graph theory_ in mathematics and the more specific category of _graph traversal_ in computer science.
+
+A network such as the Lightning Network can be represented as a mathematical construct called a _graph_, where _nodes_ are connected to each other by _edges_ (equivalent to the payment channels). The Lightning Network forms a _directed graph_ because the nodes are linked _asymmetrically_, since the channel capacity is split between the two channel partners and the payment capacity is different in each direction.
+
+==== Uncertainty of balances
+
+If we knew the exact channel balances of every channel, we could easily compute one or more payment paths using any of the standard path finding algorithms taught in good computer science programs. But we don't know the channel balances, we only know the aggregate channel capacity, which is advertised by nodes in channel announcements. In order for a payment to succeed, there must be adequate balance on the sending side of the channel. If we don't know how the capacity is distributed between the channel partners, we don't know if there is enough balance in the direction we are trying to send the payment.
```

Comment:
if the uncertainty is expressed properly we can aply the same algorithms as before. 

for single paths dijkstra with `-log(P(X>= amt))` as edges weights works just fine and the proper generalization is `min cost flows` with the same cost function. 

the same argument that you present holds. if we did not have uncertainty we could solve single path with disktra and the fee function as weights or min cost flow for cheapest multi part payment with fee funciton as weight. 

in reality we do of course a combination of costs from uncertainty and fees and in practice many implementations just do single parts (even in the mpp case but I am certain that this will change)